In [123]:
import numpy as np
import pandas as pd
import math

Problem 1: Write a program which will find all such numbers which are symetric (such as 11, 121, 1221,...) between 10 and 2000 (both included). The numbers obtained should be printed in a comma- separated sequence on a single line.

In [1]:
def is_symetric(num):
    num_list=list(str(num))
    reverse_list=num_list[::-1]
    if num_list== reverse_list:
        return True
    else:
        return False

In [2]:
symetric_list=[]
for i in range(10,2001):
    if is_symetric(i):
        symetric_list.append(i)
print('All such numbers which are symetric between 10 and 2000 (both included) are: {}'.format(str(symetric_list).strip("[]")))

All such numbers which are symetric between 10 and 2000 (both included) are: 11, 22, 33, 44, 55, 66, 77, 88, 99, 101, 111, 121, 131, 141, 151, 161, 171, 181, 191, 202, 212, 222, 232, 242, 252, 262, 272, 282, 292, 303, 313, 323, 333, 343, 353, 363, 373, 383, 393, 404, 414, 424, 434, 444, 454, 464, 474, 484, 494, 505, 515, 525, 535, 545, 555, 565, 575, 585, 595, 606, 616, 626, 636, 646, 656, 666, 676, 686, 696, 707, 717, 727, 737, 747, 757, 767, 777, 787, 797, 808, 818, 828, 838, 848, 858, 868, 878, 888, 898, 909, 919, 929, 939, 949, 959, 969, 979, 989, 999, 1001, 1111, 1221, 1331, 1441, 1551, 1661, 1771, 1881, 1991


Problem 2: Write a program which will find all such numbers which include at least 3 duplicated digits (such as 111, 1222,...) between 100 and 2000 (both included). The numbers obtained should be printed in a comma-separated sequence on a single line.


In [135]:
def is_3_duplicate(num):
    num_list=list(str(num))
    occurence_list=[num_list.count(i) for i in set(num_list)]
    greater_than3_list=[e>2 for e in occurence_list]
    if sum(greater_than3_list)> 0:
        return True
    else:
        return False
        

In [159]:
duplicate_numbers=[]
for i in range(100,2001):
    if is_3_duplicate(i):
        duplicate_numbers.append(i)

In [164]:
print('The at least 3 duplicated digits numbers are: {}'.format(str(duplicate_numbers).strip("[]")))

The at least 3 duplicated digits numbers are: 111, 222, 333, 444, 555, 666, 777, 888, 999, 1000, 1011, 1101, 1110, 1111, 1112, 1113, 1114, 1115, 1116, 1117, 1118, 1119, 1121, 1131, 1141, 1151, 1161, 1171, 1181, 1191, 1211, 1222, 1311, 1333, 1411, 1444, 1511, 1555, 1611, 1666, 1711, 1777, 1811, 1888, 1911, 1999, 2000


Problem 3: Write a program which will receive an integer larger than 0 and respond number of digits of that
integer.

In [165]:
def num_digits(number):
    if number < 0 :
        print('Please input number greater than 0')
    elif number != int(number):
        print('Please input integer')
    else:
        print(int(math.log10(number))+1)
        

In [166]:
num_digits(1235)
num_digits(-135)
num_digits(3.5)

4
Please input number greater than 0
Please input integer


Problem 4: Write a program which will receive an integer larger than 0 and present that integer in binary format.


In [26]:
def represent(number):
    new_list=[]
    while number >0:
        a=int(math.log2(number))
        new_list.append(a)
        number=number-2**a
    final_list=[10**t for t in new_list]
    print(sum(final_list))

In [30]:
represent(1456)
represent(135)

10110110000
10000111


Problem 5: Write a program which will receive an integer larger than 100 and find all divisors of that integer

In [171]:
def find_divisor(n):
    list_divisors=[n]
    for i in range(1,int(n/2)+1):
        if n%i==0:
            list_divisors.append(i)
    list_divisors=sorted(list_divisors)
    print('The divisors are {}'.format(str(list_divisors).strip("[]")))

In [172]:
find_divisor(145896)

The divisors are 1, 2, 3, 4, 6, 8, 12, 24, 6079, 12158, 18237, 24316, 36474, 48632, 72948, 145896


Problem 6: Write a program which will find all primes between 1000 and 3000 (both included). The numbers obtained should be printed in a comma-separated sequence on a single line.

In [75]:
def find_number_divisor(n):
    list_divisors=[n]
    for i in range(1,int(n/2)+1):
        if n%i==0:
            list_divisors.append(i)
    return len(list_divisors)

In [77]:
prime_list=[]
for i in range(1000,3001):
    if find_number_divisor(i)==2:
        prime_list.append(i)

In [173]:
print('All primes between 1000 and 3000 (both included) are {}'.format(str(prime_list).strip("[]")))

All primes between 1000 and 3000 (both included) are 1009, 1013, 1019, 1021, 1031, 1033, 1039, 1049, 1051, 1061, 1063, 1069, 1087, 1091, 1093, 1097, 1103, 1109, 1117, 1123, 1129, 1151, 1153, 1163, 1171, 1181, 1187, 1193, 1201, 1213, 1217, 1223, 1229, 1231, 1237, 1249, 1259, 1277, 1279, 1283, 1289, 1291, 1297, 1301, 1303, 1307, 1319, 1321, 1327, 1361, 1367, 1373, 1381, 1399, 1409, 1423, 1427, 1429, 1433, 1439, 1447, 1451, 1453, 1459, 1471, 1481, 1483, 1487, 1489, 1493, 1499, 1511, 1523, 1531, 1543, 1549, 1553, 1559, 1567, 1571, 1579, 1583, 1597, 1601, 1607, 1609, 1613, 1619, 1621, 1627, 1637, 1657, 1663, 1667, 1669, 1693, 1697, 1699, 1709, 1721, 1723, 1733, 1741, 1747, 1753, 1759, 1777, 1783, 1787, 1789, 1801, 1811, 1823, 1831, 1847, 1861, 1867, 1871, 1873, 1877, 1879, 1889, 1901, 1907, 1913, 1931, 1933, 1949, 1951, 1973, 1979, 1987, 1993, 1997, 1999, 2003, 2011, 2017, 2027, 2029, 2039, 2053, 2063, 2069, 2081, 2083, 2087, 2089, 2099, 2111, 2113, 2129, 2131, 2137, 2141, 2143, 2153, 2161,

Problem 7: Write a program which will receive an integer larger than 0 and respond all perfect numbers smaller than that integer. A perfect number is sum of its divisors except itself, such as 6 = 1+2+3. The numbers obtained should be printed in a comma-separated sequence on a single line.

In [177]:
def find_divisor(n):
    list_divisors=[n]
    for i in range(1,int(n/2)+1):
        if n%i==0:
            list_divisors.append(i)
    return list_divisors

In [178]:
def is_perfect(num):
    if sum(find_divisor(num))==2*num:
        return True
    else:
        return False

In [179]:
def search_perfect_number(num):
    perfect_number_list=[]
    for i in range(1,num):
        if is_perfect(i):
            perfect_number_list.append(i)
    return perfect_number_list
            

In [181]:
search_perfect_number(10000)

[6, 28, 496, 8128]

In [182]:
print('All perfect numbers smaller than input integer are: {}'.format(str(search_perfect_number(10000)).strip("[]")))

All perfect numbers smaller than input integer are: 6, 28, 496, 8128


Problem 8: With a given integral number n, write a program to generate a dictionary that contains (i, i*i*i) such that is an integral number between 1 and n (both included). and then the program should print the dictionary. 

In [105]:
def create_dict(n):
    new_dict={}
    for i in range(1,n+1):
        new_dict[i]=i*i*i
    return new_dict

In [109]:
create_dict(7)

{1: 1, 2: 8, 3: 27, 4: 64, 5: 125, 6: 216, 7: 343}

Problem 9: Write a program which accepts a sequence of comma separated 4 digit binary numbers as its input and then check whether they are divisible by 3 or not. The numbers that are divisible by 3 are to be printed in a comma separated sequence. Example: 0100,0011,1010,1101

In [2]:
items = []
num = [x for x in input().split(',')]
for p in num:
    x = int(p, 2)
    if not x%3:
        items.append(p)
print(','.join(items))

0100,0011,1010,1101
0011


Problem 10: Write a program which will receive an integer larger than 0 and respond if that integer is sum of
three distinct squared numbers, such as 38 = 2^2 + 3^2 + 5^2 .

In [193]:
#Create the function to check whether that integer is sum of two distinct squared numbers

def is_sum_of_2_squared(num):
    new_list=[]
    for i in range(1,int(np.sqrt(num))+1):
        if (np.sqrt(num-i**2)==int(np.sqrt(num-i**2)) and np.sqrt(num-i**2)!=i and np.sqrt(num-i**2)>0):
            new_list.append(i)
    if len(new_list)>=1:
        return True
    else:
        return False  

In [201]:
#Create the function to check whether that integer is sum of three distinct squared numbers
def is_sum_of_3_squared(num):
    new_list=[]
    for i in range(1,int(np.sqrt(num))+1):
        if (is_sum_of_2_squared(num-i**2)):
            new_list.append(i)
    if len(new_list)>=1:
        print('This number is sum of three distinct squared numbers')
    else:
        print('This number is not sum of three distinct squared numbers') 

In [203]:
is_sum_of_3_squared(38)

This number is sum of three distinct squared numbers


In [204]:
is_sum_of_3_squared(39)

This number is not sum of three distinct squared numbers


In [205]:
is_sum_of_3_squared(1090)

This number is sum of three distinct squared numbers
